x-pack/filebeat/input/entityanalytics/provider/okta/okta.go

```diff
@@ -394,18 +407,155 @@ func (p *oktaInput) doFetch(ctx context.Context, state *stateStore, fullSync boo
 	// have a complete set from that timestamp.
 	query = url.Values{}
 	query.Add("search", fmt.Sprintf(`lastUpdated ge "%s" and status pr`, lastUpdated.Format(okta.ISO8601)))
-	state.next = query.Encode()
+	state.nextUsers = query.Encode()
 
 	p.logger.Debugf("received %d users from API", len(users))
 	return users, nil
 }
 
-func grow(u []*User, n int) []*User {
-	if len(u)+n <= cap(u) {
-		return u
+// doFetchDevices handles fetching device and associated user identities from Okta.
+// If fullSync is true, then any existing deltaLink will be ignored, forcing a full
+// synchronization from Okta.
+// Returns a set of modified devices by ID.
+func (p *oktaInput) doFetchDevices(ctx context.Context, state *stateStore, fullSync bool) ([]*Device, error) {
+	if !p.cfg.WantDevices {
+		return nil, nil
+	}
+
+	var (
+		deviceQuery   url.Values
+		userQueryInit url.Values
+		err           error
+	)
+
+	// Get user changes.
+	if !fullSync && state.nextDevices != "" {
+		deviceQuery, err = url.ParseQuery(state.nextDevices)
+		if err != nil {
+			p.logger.Warnf("failed to parse next query: %v", err)
```

Comment:
What's the reason behind a Warnf and not an Errorf ? Do we expect this error to happen often ?

Reply:
There's not much to be done in the case of the error and it is partially self-healing, so I don't see the point in logging an error. This is essentially existing code though.